conda_build/pypi.py

```diff
@@ -419,6 +370,67 @@ def main(args, parser):
     print("Done")
 
 
+def get_download_data(args, client, package, version, is_url):
+    data = client.release_data(package, version) if not is_url else None
+    urls = client.release_urls(package, version) if not is_url else [package]
+    if not is_url and not args.all_urls:
+        # Try to find source urls
+        urls = [url for url in urls if url['python_version'] == 'source']
+    if not urls:
+        if 'download_url' in data:
+            urls = [defaultdict(str, {'url': data['download_url']})]
+            if not urls[0]['url']:
+                # The package doesn't have a url, or maybe it only has a wheel.
+                sys.exit("Error: Could not build recipe for %s. "
+                    "Could not find any valid urls." % package)
+            U = parse_url(urls[0]['url'])
+            if not U.path:
+                sys.exit("Error: Could not parse url for %s: %s" %
+                    (package, U))
+            urls[0]['filename'] = U.path.rsplit('/')[-1]
+            fragment = U.fragment or ''
+            if fragment.startswith('md5='):
+                md5 = fragment[len('md5='):]
+            else:
+                md5 = ''
+        else:
+            sys.exit("Error: No source urls found for %s" % package)
+    if len(urls) > 1 and not args.noprompt:
+        print("More than one source version is available for %s:" %
+                package)
+        if args.manual_url:
+            for i, url in enumerate(urls):
+                print("%d: %s (%s) %s" % (i, url['url'],
+                        human_bytes(url['size']), url['comment_text']))
+            n = int(input("which version should i use? "))
+        else:
+            print("Using the one with the least source size")
+            print("use --manual-url to override this behavior")
+            min_siz, n = min([(url['size'], i)
+                                for (i, url) in enumerate(urls)])
+    else:
+        n = 0
+
+    if not is_url:
+        print("Using url %s (%s) for %s." % (urls[n]['url'],
+            human_bytes(urls[n]['size'] or 0), package))
+        pypiurl = urls[n]['url']
+        md5 = urls[n]['md5_digest']
+        filename = urls[n]['filename']
+    else:
+        print("Using url %s" % package)
+        pypiurl = package
+        U = parse_url(package)
+        if U.fragment and U.fragment.startswith('md5='):
+            md5 = U.fragment[len('md5='):]
+        else:
+            md5 = ''
+        # TODO: 'package' won't work with unpack()
+        filename = U.path.rsplit('/', 1)[-1] or 'package'
+
+    return (data, pypiurl, filename, md5)
```

Comment:
Won't be as convenient as simply running `d.update(get_download_data)` because not everything returned is a part of dictionary `d`. See https://github.com/conda/conda-build/pull/449/files#diff-06936d397c05412b6dd38c5393523f72R328